### PR TITLE
Fix e2e network policy tests

### LINKED ISSUE #71728: Unable to run test "should enforce policy based on PodSelector [Feature:NetworkPolicy]"
```
16:14:39 • Failure in Spec Setup (BeforeEach) [28.584 seconds]
16:14:39 [sig-network] NetworkPolicy
16:14:39 /home/vagrant/go/src/github.com/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/framework.go:22
16:14:39   NetworkPolicy between server and client
16:14:39   /home/vagrant/go/src/github.com/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/network_policy.go:47
16:14:39     should enforce policy based on PodSelector [Feature:NetworkPolicy] [BeforeEach]
16:14:39     /home/vagrant/go/src/github.com/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/network_policy.go:87
16:14:39 
16:14:39     Dec  4 15:14:27.603: Pod client-can-connect-80 should be able to connect to service svc-server, but was not able to connect.
16:14:39     Pod logs:
16:14:39     /bin/sh: curl: not found
16:14:39     /bin/sh: curl: not found
16:14:39     /bin/sh: curl: not found
16:14:39     /bin/sh: curl: not found
16:14:39     /bin/sh: curl: not found
16:14:39     
16:14:39     
16:14:39      Current NetworkPolicies:
16:14:39     	[]
16:14:39     
16:14:39      Pods:
16:14:39     	[Pod: server, Status: &PodStatus{Phase:Running,Conditions:[{Initialized True 0001-01-01 00:00:00 +0000 UTC 2018-12-04 15:14:09 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2018-12-04 15:14:17 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2018-12-04 15:14:17 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2018-12-04 15:14:09 +0000 UTC  }],Message:,Reason:,HostIP:192.168.36.12,PodIP:10.10.1.237,StartTime:2018-12-04 15:14:09 +0000 UTC,ContainerStatuses:[{server-container-80 {nil ContainerStateRunning{StartedAt:2018-12-04 15:14:13 +0000 UTC,} nil} {nil nil nil} true 0 gcr.io/kubernetes-e2e-test-images/porter:1.0 docker-pullable://gcr.io/kubernetes-e2e-test-images/porter@sha256:d6389405e453950618ae7749d9eee388f0eb32e0328a7e6583c41433aa5f2a77 docker://d4db04bb6146e4e80200f1ba142625430ac5170dbb5fb5a7a0fb3956ed39a142} {server-container-81 {nil &ContainerStateRunning{StartedAt:2018-12-04 15:14:13 +0000 UTC,} nil} {nil nil nil} true 0 gcr.io/kubernetes-e2e-test-images/porter:1.0 docker-pullable://gcr.io/kubernetes-e2e-test-images/porter@sha256:d6389405e453950618ae7749d9eee388f0eb32e0328a7e6583c41433aa5f2a77 docker://e3a08ed763f4d8e852cd7758c03e7bdc281a42a3e18d5252e4ddcd311a75da7e}],QOSClass:BestEffort,InitContainerStatuses:[],NominatedNodeName:,}
16:14:39      Pod: server, Status: &PodStatus{Phase:Running,Conditions:[{Initialized True 0001-01-01 00:00:00 +0000 UTC 2018-12-04 15:14:09 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2018-12-04 15:14:17 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2018-12-04 15:14:17 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2018-12-04 15:14:09 +0000 UTC  }],Message:,Reason:,HostIP:192.168.36.12,PodIP:10.10.1.237,StartTime:2018-12-04 15:14:09 +0000 UTC,ContainerStatuses:[{server-container-80 {nil ContainerStateRunning{StartedAt:2018-12-04 15:14:13 +0000 UTC,} nil} {nil nil nil} true 0 gcr.io/kubernetes-e2e-test-images/porter:1.0 docker-pullable://gcr.io/kubernetes-e2e-test-images/porter@sha256:d6389405e453950618ae7749d9eee388f0eb32e0328a7e6583c41433aa5f2a77 docker://d4db04bb6146e4e80200f1ba142625430ac5170dbb5fb5a7a0fb3956ed39a142} {server-container-81 {nil &ContainerStateRunning{StartedAt:2018-12-04 15:14:13 +0000 UTC,} nil} {nil nil nil} true 0 gcr.io/kubernetes-e2e-test-images/porter:1.0 docker-pullable://gcr.io/kubernetes-e2e-test-images/porter@sha256:d6389405e453950618ae7749d9eee388f0eb32e0328a7e6583c41433aa5f2a77 docker://e3a08ed763f4d8e852cd7758c03e7bdc281a42a3e18d5252e4ddcd311a75da7e}],QOSClass:BestEffort,InitContainerStatuses:[],NominatedNodeName:,}
16:14:39     ]
```

As far as I know the busybox image doesn't contain `curl`, only `wget` so I don't understand how the PR that changed from `wget` to `curl` has passed the CI https://github.com/kubernetes/kubernetes/pull/69571

**What happened**:

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

```
git clone https://github.com/kubernetes/kubernetes.git -b v1.13.0 --depth 1
cd kubernetes
make ginkgo
make WHAT='test/e2e/e2e.test'

export KUBERNETES_PROVIDER=vagrant
export KUBECTL_PATH=/usr/bin/kubectl
export KUBE_MASTER=192.168.36.11
export KUBE_MASTER_IP=192.168.36.11
export KUBE_MASTER_URL="https://192.168.36.11:6443"

go run hack/e2e.go --test --test_args="--ginkgo.focus=NetworkPolicy --e2e-verify-service-account=false --host ${KUBE_MASTER_URL}"
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/kind bug

cc @bowei

### LINKED ISSUE #74289: e2e tests NetworkPolicies are failing
**What happened**:

I wanted to run the e2e network policy tests with: `kubetest --provider=skeleton --test --test_args="--ginkgo.focus=\[Feature:NetworkPolicy\]"` and I noticed that all tests are failing (I actually don't know why nobody has noticed this before).  Looking into the logs I saw the following errors:

```bash
    Pod logs:
    /bin/sh: curl: not found
    /bin/sh: curl: not found
    /bin/sh: curl: not found
    /bin/sh: curl: not found
    /bin/sh: curl: not found
...    
```

which are correct because the tests using `busybox:1.29` as "client" which doesn't contain `curl`.

**What you expected to happen**:

The e2e tests for network policies should actually pass.

**How to reproduce it (as minimally and precisely as possible)**:

Just run: `kubetest --provider=skeleton --test --test_args="--ginkgo.focus=\[Feature:NetworkPolicy\]"`

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):

```bash
kubectl version                                                                                             
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", BuildDate:"2019-02-04T04:48:03Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.2", GitCommit:"cff46ab41ff0bb44d8584413b598ad8360ec1def", GitTreeState:"clean", BuildDate:"2019-01-15T23:29:42Z", GoVersion:"go1.11.4", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: kind
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by johscheuer**
**What type of PR is this?**
/kind failing-test

**What this PR does / why we need it**:

This PR uses wget instead of curl (because curl is missing in the busybox image) to make the network policy tests pass.

**Which issue(s) this PR fixes**:

Fixes #74289
Fixes #71728

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by danwinship** [comment]
also
Fixes #71728
(but I think github will only pick that up if it's in the initial comment, so can you edit that comment to add it there?)

**Comment by danwinship** [comment]
As noted in #71728, this was done to make the tests work on Windows; the catch is that they forgot to change the test to use an image that actually contains curl. So the fix is to change the image used by the test, not to revert back to wget.

**Comment by johscheuer** [comment]
I would choose `centos:7` for the tests since centos has curl included (unlike ubuntu) or are there any arguments against centos?

**Comment by johscheuer** [comment]
From my point of view a revert back to wget is still a valid option since https://github.com/kubernetes/kubernetes/pull/69571/files#diff-b7cdbd8f1585bcdaf1faadb3b17ac950 is only executed in the container and not on the host so it should be irrelevant if you are running Linux/MacOS/Windows etc.. The tests in https://github.com/kubernetes/kubernetes/pull/69571/files#diff-cb4c40bc4199d52e8b5e9e7f26e84d91 are the only tests that are actually performed on the host and should use `curl` (but this file wasn't changed by my initial commit).

**Comment by danwinship** [comment]
Yes, I think you're right. Sorry for the confusion. Can you revert back to your original fix?

also, ping @bclau who made the original change

**Comment by johscheuer** [comment]
Reverted back to the original changes with `wget`

**Comment by bclau** [comment]
> 
> 
> From my point of view a revert back to wget is still a valid option since https://github.com/kubernetes/kubernetes/pull/69571/files#diff-b7cdbd8f1585bcdaf1faadb3b17ac950 is only executed in the container and not on the host so it should be irrelevant if you are running Linux/MacOS/Windows etc.. The tests in https://github.com/kubernetes/kubernetes/pull/69571/files#diff-cb4c40bc4199d52e8b5e9e7f26e84d91 are the only tests that are actually performed on the host and should use `curl` (but this file wasn't changed by my initial commit).

The container still has an OS, and in our case the container's OS is Windows. So, this will break the tests on Windows once again.

I apologize for not checking if it would still be working on Linux nodes, I was expecting the CIs to do that for me.

I'd still go with the curl option, but clearly we should change the image. There already is an image which has curl, so we could use that. Looking into it.

**Comment by danwinship** [comment]
> The container still has an OS, and in our case the container's OS is Windows. So, this will break the tests on Windows once again.

But the image is busybox... it doesn't matter if it's busybox on Linux or busybox on Windows, it ought to have a "wget" binary in it.

**Comment by danwinship** [comment]
@johscheuer can you squash everything back down to 1 commit and force push?
Also, what's the `-O` for? Just debuggability?

**Comment by johscheuer** [comment]
The `-O -` was in the original file: https://github.com/kubernetes/kubernetes/pull/69571/files#diff-b7cdbd8f1585bcdaf1faadb3b17ac950 actually I don't think we need it there.

**Comment by johscheuer** [comment]
@bclau or do you overwrite the `imageutils.BusyBox` in some kind of a fork? I would assume that if you actually want to run these network policy tests with a Windows container that you have some kind of an if statement in the code to use either the linux container image or the windows container image (actually I don't know if the e2e are currently built to run on Windows Servers)

**Comment by bclau** [comment]
> 
> 
> > The container still has an OS, and in our case the container's OS is Windows. So, this will break the tests on Windows once again.
> 
> But the image is busybox... it doesn't matter if it's busybox on Linux or busybox on Windows, it ought to have a "wget" binary in it.

It does have wget, but its functionality is limited. For example, there doesn't seem to have any -T arguments:

    kubectl exec -n network-policy client-can-connect-80 -- wget --help
    BusyBox v1.29.0.git (2018-10-29 11:20:48 PDT) multi-call binary

    Usage: wget [-c|--continue] [--spider] [-q|--quiet] [-O|--output-document FILE]
            [--header 'header: value'] [-Y|--proxy on/off] [-P DIR]
            [-S|--server-response] [-U|--user-agent AGENT] URL...

    Retrieve files via HTTP or FTP

            --spider        Only check URL existence: $? is 0 if exists
            -c              Continue retrieval of aborted transfer
            -q              Quiet
            -P DIR          Save to DIR (default .)
            -S              Show server response
            -O FILE         Save to FILE ('-' for stdout)
            -U STR          Use STR for User-Agent header
            -Y on/off       Use proxy

Keep in mind that there are quite a few significant differences between Windows and Linux, and those differences are also reflected in busybox binary's functionality. One other difference that comes to mind would be the nc binary, which, again, is quite limited in the busybox's implementation.

Alternatively, we could remove the -T argument and increase the number of iteration in the loop to avoid flakiness (in case the container started before the network could be fully set up for the container).

> 
> 
> @johscheuer can you squash everything back down to 1 commit and force push?
> Also, what's the `-O` for? Just debuggability?

-O - will output the content to stdout instead of index.html.

> 
> 
> The `-O -` was in the original file: https://github.com/kubernetes/kubernetes/pull/69571/files#diff-b7cdbd8f1585bcdaf1faadb3b17ac950 actually I don't think we need it there.

It isn't required, but the test collects the pod's logs and logs them in the test. It could be useful to determine what didn't work. So, for debugability's purposes, I'd still add the "-O -".

**Comment by johscheuer** [comment]
Could you tell us which Windows busybox container you are using? I would then take a look to find the command that can be used by the Linux and Windows busybox to make the tests run on Linux and Windows. Probably one solution could be to use `nc -vz -w 1 svc port` ?

**Comment by johscheuer** [comment]
At least on Linux this worked: 

```bash
Ran 6 of 3004 Specs in 361.312 seconds
SUCCESS! -- 6 Passed | 0 Failed | 0 Pending | 2998 Skipped PASS
```

could you cross check that the Windows busybox also contains `nc`:

```bash
BusyBox v1.29.3 (2018-12-24 21:25:20 UTC) multi-call binary.

Usage: nc [OPTIONS] HOST PORT  - connect
nc [OPTIONS] -l -p PORT [HOST] [PORT]  - listen

	-e PROG	Run PROG after connect (must be last)
	-l	Listen mode, for inbound connects
	-lk	With -e, provides persistent server
	-p PORT	Local port
	-s ADDR	Local address
	-w SEC	Timeout for connects and final net reads
	-i SEC	Delay interval for lines sent
	-n	Don't do DNS resolution
	-u	UDP mode
	-v	Verbose
	-o FILE	Hex dump traffic
	-z	Zero-I/O mode (scanning)
```

**Comment by bclau** [comment]
Just so you know, the Windows busybox nc is extremely limited:

    PS C:\Users\Administrator\Documents> .\busybox64.exe nc -h
    C:\Users\Administrator\Documents\busybox64.exe: unknown option -- h
    BusyBox v1.31.0-FRP-2962-g9aae70405 (2019-02-12 08:57:53 GMT) multi-call binary

    Usage: nc [-l] [-p PORT] [IPADDR PORT]

    Open a pipe to IP:PORT

            -l      Listen mode, for inbound connects
            -p PORT Local port

This is a known issue, and I've tried to build a more feature-complete version of busybox with all the necessary nc features, but it cannot be built on Windows at the moment: https://github.com/rmyorston/busybox-w32/issues/117

Instead, I could include a different, more feature complete nc.exe inside the busybox image, one that we're already using in the hostexec image: https://github.com/kubernetes-sigs/windows-testing/blob/f96cb4cc6d71cc5b9782306ed89cd93d1363ccd8/images/hostexec/Dockerfile#L17

The busybox image we're using is ``e2eteam/busybox:1.29`` btw.

**Comment by johscheuer** [comment]
Adding the more feature complete `nc` to be busybox image for the Windows container sound to me like a good solution (and we just use `nc` instead of `curl`/`wget` in the tests).

If you want I can take a look at the busybox image for Windows and add `nc`

**Comment by bclau** [comment]
I was already checking out how it works with nc. Seems fine to me. I've already made a PR regarding the nc.exe in the windows busybox image.

**Comment by johscheuer** [comment]
Nice ! So when I understand you correctly we can merge the changes in this PR as they are?

File: test/e2e/network/network_policy.go [hunk 1]
```diff
@@ -17,7 +17,7 @@ limitations under the License.
 package network
 
 import (
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (bclau) [inline test/e2e/network/network_policy.go]: fyi, unrelated change that doesn't really do anything.

**Comment by bclau** [comment]
Yep, works for windows, so:

/lgtm

**Comment by bowei** [comment]
/approve
/lgtm

thanks for the fix

**Comment by lzang** [comment]
@johscheuer, could you please back port this change to 1.13 as well? Thanks!

**Comment by johscheuer** [comment]
If you tell me how :)

**Comment by lzang** [comment]
@johscheuer OK, I create a PR here to port your changes. 
https://github.com/kubernetes/kubernetes/pull/75219

**Pull request merged**